README.md
# anyvpn - a tool to connect to the Cisco AnyConnect VPN

Getting your corporate password from your password
manager whenever you connect to the VPN is a pain.
This tool can help you make this less painful and
hopefully encourage users to use a password manager.

At this time, this script supports the use of the
macOS keychain, LastPass, and 1Password.  For use with
LastPass, you need to install the `lpass` command-line
utility (available from e.g.,
[here](https://github.com/lastpass/lastpass-cli)); for
use with 1Password, you need to install the `op`
command-line utility (available from e.g.,
[here](https://support.1password.com/command-line-getting-started/)).



Installation
============

To install the command and manual page somewhere
convenient, run `make install`; the Makefile defaults
to '/usr/local' but you can change the PREFIX:

```
$ make PREFIX=~ install
```

Documentation
=============

Please see the manual page for all details:


```
NAME
     anyvpn - connect to a Cisco AnyConnect VPN

SYNOPSIS
     anyvpn [-Vhlv] [-m method] [-p app] [-s site] [-u user] on|off|sites|vars

DESCRIPTION
     anyvpn is a simple script that allows you to connect to a Cisco AnyConnect
     VPN via the commandline by supplying your password and 2FA method so as to
     minimize interactions with the process.

OPTIONS
     The following options are supported by anyvpn:

     -V		Print the current version information and exit.

     -h		Display help and exit.

     -l		Log out of the password manager after fetching the password.

     -m method	Use this 2FA method.  Valid options are 'push' or a Duo 2FA
		code.  Defaults to 'push'.

     -p app	Use this password manager app.	Currently supported: keychain,
		lastpass, onepass.  If not specified, anyvpn will use
		'lastpass'.

     -s site	Connect to this site.  Defaults to the value of the VPN_SITE
		environment variable or, if that is unset, the first site
		condigured in AnyConnect.

     -u user	Set the username to use when connecting to the VPN.  This is
		similar to but overrides setting the value via the VPN_USER
		environment variable.  (This option exists to allow the user to
		use a different username ad-hoc, but set a default different
		from the local username for common usage.)

     -v		Be verbose.  Can be specified multiple times.

COMMANDS
     The following commands are supported by anyvpn:

     on	     Connect to the given VPN site.

     off     Disconnect from the VPN.

     sites   List the available VPN sites you can connect to.

     vars    Print out the value of all relevant environment variables used by
	     anyvpn.

     Any other arguments are passed through to the AnyConnect VPN command.

DETAILS
     In order to connect to a VPN via Cisco AnyConnect, you may need to provide
     your password as well as a second factor.	Retrieving your password from a
     password manager can be cumbersome, since AnyConnect does not integrate
     with e.g., LastPass, leading to many people choosing to use an easy to
     remember VPN password.

     anyvpn provides the necessary glue to make it easier to connect to the VPN
     without having to unlock your password manager and copy and paste your
     password.

PASSWORD MANAGER CONSIDERATIONS
     In order to retrieve your password from your password manager, and unless
     you are using the macOS keychain (-p keychain), anyvpn needs to begin a
     session with the password manager or reuse an existing one.

     The behavior with respect to new sessions is as outlined below:

     o	 If an existing session is found ("lpass status -q" for lpass(1); a
	 valid session token in the ONEPASS_SESSION environment variable for
	 op(1)), anyvpn will reuse that session.

     o	 If no existing session is found, anyvpn will initiate a new session.

     Since a new session for op(1) is exposed only via an in-memory token,
     termination of anyvpn effectively makes that session unavailable for any
     other processes, even if it still remains technically active, which is why
     anyvpn explicitly invalidates it if it originally started it.

     On the other hand, a lpass(1) session initiated by anyvpn will remain
     active after the script terminates.  Such a valid password manager session
     may then allow other processes to access secrets from that password
     manager.  This is by design and can in fact be useful when using other
     tools that may require a password to be pulled from the password manager.

     To ensure anyvpn invalidates the newly started session, pass the -f flag.

EXIT STATUS
     The anyvpn utility will exit with a value of 255 if it encounters any
     problems prior to executing the AnyConnect 'vpn' command.	Otherwise,
     anyvpn will return with the exit value of the AnyConnect 'vpn' command.

EXAMPLES
     The following examples illustrate common usage of the anyvpn utility.

     To add a new password to the "VPN" keychain service and then use anyvpn to
     connect to the VPN using that password from the keychain:

	   security add-generic-password -a ${USER} -s VPN -w '<password>'
	   anyvpn -k keychain on

     To connect to the VPN using the password stored under the "Yahoo" login
     entry in your LastPass password manager for the "jschauma@yahoo.com"
     account:

	   export LASTPASS_LOGIN="jschauma@yahoo.com"
	   export PM_VPN_ENTRY="Yahoo"
	   anyvpn on

     To connect to the VPN using the password stored under the "Yahoo" login
     entry in your 1Password password manager for the "jschauma@yahoo.com"
     account in your company's team vault "mycorp.1password.com":

	   export USER="jschauma"
	   export DOMAIN="yahoo.com"
	   export PM_VPN_ENTRY="Yahoo"
	   export ONEPASS_ADDRESS="mycorp.1password.com"
	   anyvpn -p onepass on

     To connect to the "NYC" VPN site using an existing 1Password session from
     your personal 1Password account:

	   export ONEPASS_SESSION="$(op signin --raw)"
	   export PM_VPN_ENTRY="Yahoo"
	   anyvpn -s NYC -p onepass on

ENVIRONMENT
     The following environment variables affect the execution of this tool:

     ANYCONNECT_PREFIX	  The path to where AnyConnect is installed.  Defaults
			  to "/opt/cisco/anyconnect".

     DOMAIN		  The domain to use when constructing your
			  LASTPASS_LOGIN or ONEPASS_LOGIN ID.

			  If unset, then anyvpn will try to identify a suitable
			  domain by looking at /etc/resolv.conf or the output of
			  the hostname(1) command.

     KEYCHAIN_VPN_ENTRY	  The name of the keychain entry ("service") for your
			  VPN login.

			  If unset, then anyvpn will use "VPN".

     LASTPASS_LOGIN	  The full LastPass user ID.  This may be e.g.,
			  "first.lass@company.tld", "${USER}@company.name" etc.

			  If unset, then anyvpn will use "${USER}@${DOMAIN}".

     LASTPASS_VPN_ENTRY	  The name of the LastPass entry for your VPN login.

			  If unset, then anyvpn will use the value of the
			  PM_VPN_ENTRY environment variable.

     LPASS_AGENT_TIMEOUT  Not directly used by anyvpn, but used by lpass(1),
			  this variable defines in seconds the validity of your
			  LastPass session.  Set this to e.g., 28800 for an 8
			  hour LastPass cache validity.

     ONEPASS_ADDRESS	  The 1Password "sign in address".  If not specified,
			  defaults to "my.1password.com".  See
			  https://is.gd/BR670l for details.

     ONEPASS_LOGIN	  The 1Password user ID.  This may be e.g.,
			  "first.lass@company.tld", "${USER}@company.name" etc.

			  If unset, then anyvpn will use "${USER}@${DOMAIN}".

     ONEPASS_SESSION	  The 1Password op(1) session token to use.  Setting
			  this in your environment allows you to sign in to your
			  1Password account in your shell, and anyvpn to
			  retrieve the password without requiring your master
			  password to be entered.

			  If unset, then anyvpn will sign in to your 1Password
			  account, prompting you for your master password,
			  retrieve the VPN password, and then invalidate the
			  session.

     ONEPASS_VPN_ENTRY	  The name of the 1Password entry for your VPN login.

			  If unset, then anyvpn will use the value of the
			  PM_VPN_ENTRY environment variable.

     PM_VPN_ENTRY	  The name of the password manager entry for your VPN
			  login.  If unset, then anyvpn will use "VPN".

			  Note: this value is overriden by either of the
			  KEYCHAIN_VPN_ENTRY, LASTPASS_VPN_ENTRY, or
			  ONEPASS_VPN_ENTRY environment variables.

     USER		  The local username, as commonly set on most unix
			  systems.  This variable is not used directly, but some
			  of the other variables described here may derive their
			  default value from this variable.

			  If unset, then anyvpn will use the output of 'id -un'.

     VPN_SITE		  The name of the VPN site to connect to.  Supported
			  values can be shown by specifying sites as the
			  argument to anyvpn.

			  If unset, then anyvpn will use the first site
			  configured in AnyConnect.

     VPN_USER		  The short name / user ID, commonly your VPN or SSO
			  login name.

			  If unset, then anyvpn will use "${USER}".

			  Note: specifying the user via the -u flag overrides
			  this value.

SEE ALSO
     lpass(1), op(1), security(1)

HISTORY
     This script was originally written by Jan Schaumann
     <jschauma@netmeister.org> in June 2020.

BUGS
     Please file bugs and feature requests via GitHub pull requests and issues
     or by emailing the author.
```


CHANGES
Version 1.4 (2022-03-16)
 * add an option to dump variables and values used
 * actually allow the DOMAIN variable to override
   built-in defaults

Version 1.3 (2022-03-10)
 * add '-u user' flag

Version 1.2 (2022-03-02)
 * add support for macOS keychain via security(1)

Version 1.1 (2022-01-28)
 * handle rare lpass status failure

Version 1.0 (2022-01-20)
 * actually make '-f' do what we think was intended

Version 0.9 (2021-06-17)
 * add '-s' to usage output
 * add '-f' to force password manager logout

Version 0.8 (2021-05-26)
 * rename to "anyvpn"

Version 0.7 (2021-04-13)
 * add support for 1Password

Version 0.6 (2021-04-08)
 * account for multiple identical lpass entries
 * detect when we're already connected / disconnected

Version 0.5 (2021-04-08)
 * actually honor VPN_USER

Version 0.4 (2021-02-17)
 * first public version

Version 0.3 (2020-07-02)
 * pass through args to the anyconnect vpn command

Version 0.2 (2020-06-11)
 * add '-s' to select site

Version 0.1 (2020-06-10)
 * first import into git


src/anyvpn.sh
#! /bin/sh
#
# This script can be used to use Cisco Anyconnect to
# connect to a VPN and have lpass(1) or op(1) provide
# the password.

set -eu

PROGNAME="${0##*/}"
VERSION="1.4"
VERBOSITY=0

# These will be determined below in setVariables()
_DOMAIN=""
_DOMAIN_SRC="default         "

_LP_LOGIN=""
_OP_LOGIN=""
_SITES=""
_VPN_SITE=""
_VPN_USER=""

_FAIL=255
_FLAGS=""
_FORCE_LOGOUT=0

_2FA="push"
_ANYCONNECT_PREFIX="${ANYCONNECT_PREFIX:-"/opt/cisco/anyconnect"}"
_ANYCONNECT="${_ANYCONNECT_PREFIX}/bin/vpn"
_KC="$(which security || true)"
_LPASS="$(which lpass || true)"
_OPASS="$(which op || true)"
_OPASS_SIGNIN_ADDRESS="${ONEPASS_ADDRESS:-"my.1password.com"}"
_PASSWORD_MANAGER="lastpass"
_VPN_PW=""

_PM_VPN="${PM_VPN_ENTRY:-"VPN"}"
_KC_VPN="${KEYCHAIN_VPN_ENTRY:="${_PM_VPN}"}"
_LP_VPN="${LASTPASS_VPN_ENTRY:-"${_PM_VPN}"}"
_OP_VPN="${ONEPASS_VPN_ENTRY:-"${_PM_VPN}"}"

###
### Functions
###

checkAnyconnectState() {
	local state="$1"

	if [ x"${state}" != x"connected" ] && [ x"${state}" != x"disconnected" ]; then
		echo "Invalid state: ${state}" >&2
		exit ${_FAIL}
	fi

	if ${_ANYCONNECT} state 2>&1 | grep -q -i "state: ${state}" >/dev/null 2>&1; then
		echo "Already ${state}." >&2
		exit ${_FAIL}
		# NOTREACHED
	fi
}

checkEnv() {
	local l found

	case "${_PASSWORD_MANAGER}" in
		lpass)
			_PASSWORD_MANAGER="lastpass"
		;;
		opass|1pass|1password)
			_PASSWORD_MANAGER="onepass"
		;;
	esac

	if [ x"${_PASSWORD_MANAGER}" = x"lastpass" ]; then
		if [ -z "${_LPASS}" ]; then
			echo "Unable to find 'lpass(1)' in your PATH." >&2
			echo "Please install the lastpass-cli via e.g., 'brew install lastpass-cli' or from" >&2
			echo "https://github.com/lastpass/lastpass-cli"
			exit ${_FAIL}
		fi
	elif [ x"${_PASSWORD_MANAGER}" = x"onepass" ]; then
		if [ -z "${_OPASS}" ]; then
			echo "Unable to find the 'op(1)' in your PATH." >&2
			echo "Please install the 1Password command-line client from" >&2
			echo "https://support.1password.com/command-line-getting-started/" >&2
			exit ${_FAIL}
		fi
	elif [ x"${_PASSWORD_MANAGER}" = x"keychain" ]; then
		if [ -z "${_KC}" ]; then
			echo "Unable to find 'security(1)' in your PATH." >&2
			exit ${_FAIL}
		fi
	else
		echo "Unsupported password manager." >&2
		exit ${_FAIL}
		# NOTREACHED
	fi

	if [ ! -x "${_ANYCONNECT}" ]; then
		echo "Unable to find '${_ANYCONNECT}'." >&2
		exit ${_FAIL}
	fi

	oIFS="${IFS}"
	IFS='
'

	found=0
	for l in ${_SITES}; do
		if [ x"${_VPN_SITE}" = x"${l}" ]; then
			found=1
			break
		fi
	done
	IFS="${oIFS}"

	if [ ${found} -ne 1 ]; then
		echo "Invalid VPN_SITE '${_VPN_SITE}'." >&2
		echo "Supported sites:" >&2
		echo "${_SITES}" >&2
		exit ${_FAIL}
		# NOTREACHED
	fi

	if [ x"${_2FA}" != x"push" ] &&
		! expr "${_2FA}" : "[0-9]*$" >/dev/null 2>&1 ; then
		echo "Invalid 2FA method '${_2FA}'." >&2
		exit ${_FAIL}
		# NOTREACHED
	fi
}

connectToVPN() {
	verbose "Connecting to '${_VPN_SITE}' as '${_VPN_USER}' with 2FA method '${_2FA}'..." 2

	${_ANYCONNECT} -s << EOF
connect "${_VPN_SITE}"
${_VPN_USER}
${_VPN_PW}
${_2FA}
EOF
}

dumpVars() {
	verbose "Dumping the values of all relevant variables we use..."

	local ac="default             "
	if [ -n "${ANYCONNECT_PREFIX:-""}" ]; then
		ac="\${ANYCONNECT_PREFIX}"
	fi

	local pm="${_LPASS}"
	local pmentry="default              "
	if [ -n "${PM_VPN_ENTRY:-""}" ]; then
		pmentry="\${PM_VPN_ENTRY}      "
	fi
	local pmval="${_PM_VPN}"
	local pmlogin="\${user}@\${domain}"
	local pmloginval
	case "${_PASSWORD_MANAGER}" in
		"lastpass")
			pm="${_LPASS}"
			pmloginval="${_LP_LOGIN}"
			if [ -n "${LASTPASS_VPN_ENTRY:-""}" ]; then
				pmentry="\${LASTPASS_VPN_ENTRY}"
				pmval="${_LP_VPN}"
			fi
			if [ -n "${LASTPASS_LOGIN:=""}" ]; then
				pmlogin="\${LASTPASS_LOGIN}"
			fi
		;;
		"onepass")
			pm="${_OPASS}"
			pmloginval="${_OP_LOGIN}"
			if [ -n "${ONEPASS_VPN_ENTRY:-""}" ]; then
				pmentry="\${ONEPASS_VPN_ENTRY} "
				pmval="${_OP_VPN}"
			fi
			if [ -n "${ONEPASS_LOGIN:=""}" ]; then
				pmlogin="\${ONEPASS_LOGIN} "
			fi
		;;
		"keychain")
			pm="${_KC}"
			pmlogin="N/A"
			pmloginval="N/A"
			if [ -n "${KEYCHAIN_VPN_ENTRY:-""}" ]; then
				pmentry="\${KEYCHAIN_VPN_ENTRY}"
				pmval="${_KC_VPN}"
			fi
		;;
	esac
	local pmo="default"
	if expr "${_FLAGS}" : ".* p" >/dev/null; then
		pmo="-p     "
	fi

	local vpnu="\${USER}    "
	if [ -n "${VPN_USER:-""}" ]; then
		vpnu="\${VPN_USER}"
	fi
	if expr "${_FLAGS}" : ".* u" >/dev/null; then
		vpnu="-u         "
	fi

	local vpns="default    "
	if [ -n "${VPN_SITE:-""}" ]; then
		vpns="\${VPN_SITE}"
	fi
	if expr "${_FLAGS}" : ".* s" >/dev/null; then
		vpns="-s         "
	fi

	local mfo="default"
	if expr "${_FLAGS}" : ".* m" >/dev/null; then
		mfo="-m     "
	fi

	cat <<EOF
Setting                      Derived from            Value used
-----------------------------------------------------------------------------
AnyConnect Path              ${ac}    "${_ANYCONNECT_PREFIX}"
Domain                       ${_DOMAIN_SRC}        "${_DOMAIN}"
MFA Option                   ${mfo}                 "${_2FA}"
Password Manager             ${pmo}                 "${_PASSWORD_MANAGER}"
Password Manager Executable                          "${pm}"
Password Manager Entry       ${pmentry}   "${pmval}"
Password Manager Login       ${pmlogin}       "${pmloginval}
VPN User                     ${vpnu}             "${_VPN_USER}"
VPN Site                     ${vpns}             "${_VPN_SITE}"
EOF
}

getPasswordFromKeychain() {
	verbose "Trying to get your VPN password from the keychain..." 2

	_VPN_PW="$(${_KC} find-generic-password -a ${USER} -s ${_KC_VPN} -w || true)"
	if [ -z "${_VPN_PW}" ]; then
		echo "Unable to retrieve password from keychain service '${_KC_VPN}'." >&2
		exit ${_FAIL}
		# NOTREACHED
	fi
}

getPasswordFromLastPass() {
	verbose "Trying to get your VPN password from LastPass..." 2

	# 'lpass status' may sometimes report being
	# "Logged in as (null)", and 'lpass status -q'
	# would return successfully, so we can't rely
	# on that and instead have to check for an
	# email address.
	if ! ${_LPASS} status | grep -q -i 'Logged in as .*@'; then
		verbose "Trying to log into LastPass..." 3
		${_LPASS} login "${_LP_LOGIN}"
	fi
	
	_VPN_PW="$(${_LPASS} show "${_LP_VPN}" --password)"
	if [ -z "${_VPN_PW}" ]; then
		echo "Got an empty password from LastPass??" >&2
		exit ${_FAIL}
		# NOTREACHED
	fi

	if expr "${_VPN_PW}" : "Multiple matches found." >/dev/null 2>&1; then
		echo "${_VPN_PW}" >&2
		echo >&2
		echo "If both are identical, you could delete one or the other." >&2
		echo "If they are different, you could rename one or the other." >&2
		echo "If you want to keep both as they are, then please set" >&2
		echo "the LASTPASS_VPN_ENTRY environment variable to the correct ID." >&2
		exit ${_FAIL}
		# NOTREACHED
	fi
}

getPasswordFromOnePass() {
	verbose "Trying to get your VPN password from 1Password..." 2

	local token="${ONEPASS_SESSION:-""}"
	local args="${_OPASS_SIGNIN_ADDRESS}"
	local signin=0

	if [ -z "${token}" ]; then
		verbose "Signing into 1Password..." 3
		# Provide the complete signin address only the
		# first time op signs in.  This will require
		# the user to provide the "Secret Key".
		if [ ! -f ~/.op/config ]; then
			args="${args} ${_OP_LOGIN}"
		fi
		token="$(${_OPASS} signin --raw "${args}")"
		signin=1
	fi

	_VPN_PW="$(${_OPASS} --session "${token}" get item "${_OP_VPN}" --fields password)"
	if [ -z "${_VPN_PW}" ]; then
		echo "Got an empty password from 1Password??" >&2
		exit ${_FAIL}
		# NOTREACHED
	fi

	if [ ${signin} -eq 1 ]; then
		verbose "Signing out of 1Password..." 3
		${_OPASS} --session "${token}" signout
	fi
}

getPasswordFromPasswordManager() {
	case "${_PASSWORD_MANAGER}" in
		keychain)
			getPasswordFromKeychain
		;;
		lastpass)
			getPasswordFromLastPass
		;;
		onepass)
			getPasswordFromOnePass
		;;
	esac

	if [ ${_FORCE_LOGOUT} -gt 0 ]; then
		logoutPass
	fi
}

listSites() {
	echo "The following values are supported for VPN_SITE:"
	echo "${_SITES}"
	exit 0
}

logoutPass() {
	verbose "Logging out of ${_PASSWORD_MANAGER}..."

	case "${_PASSWORD_MANAGER}" in
		lastpass)
			${_LPASS} logout -f
		;;
		# For 1Password, we explicitly log in, or require a login
		# session to already exist.  If that already exists, we
		# don't want to log the user out here, so noop.
	esac
}

setVariables() {
	local user

	if [ -n "${DOMAIN:-""}" ]; then
		_DOMAIN="${DOMAIN}"
		_DOMAIN_SRC="\${DOMAIN}     "
	elif [ -z "${_DOMAIN:-""}" ]; then
		if [ -f "/etc/resolv.conf" ]; then
			_DOMAIN="$(sed -n -E 's/^(search|domain) ([^ ]+).*/\2/p' /etc/resolv.conf)"
			_DOMAIN_SRC="/etc/resolv.conf"
	else
		# 'hostname -f' is not portable
			_DOMAIN="$(hostname)"
			_DOMAIN_SRC="hostname(1)     "
		fi
	fi

	if [ -n "${USER:-""}" ]; then
		user="${USER}"
	else
		user="$(id -un)"
	fi

	_LP_LOGIN="${LASTPASS_LOGIN:-"${user}@${_DOMAIN}"}"
	_OP_LOGIN="${ONEPASS_LOGIN:-"${user}@${_DOMAIN}"}"

	_VPN_USER="${VPN_USER:-"${user}"}"

	_SITES="$(sed -n -e 's/.*<HostName>\(.*\)<\/HostName>.*/\1/p' ${_ANYCONNECT_PREFIX}/profile/*.xml)"

	if [ -n "${VPN_SITE:-""}" ]; then
		_VPN_SITE="${VPN_SITE}"
	else
		_VPN_SITE="$(echo "${_SITES}" | head -1)"
	fi
}

usage() {
	cat <<EOH
Usage: ${PROGNAME} [-Vhlv] [-m method] [-p app] [-s site] [-u user] on|off|sites|vars
	-V         print version number and exit
	-h         print this help and exit
	-l         log out of the password manager after fetching the password
	-m method  use this 2FA method (default: push)
	-p app     use this password manager (keychain, lastpass, onepass; default: lastpass)
	-s site    connect to this site (default: ${_VPN_SITE})
	-u user    use this username to connect to the VPN (default: ${_VPN_USER})
	-v         be verbose
EOH
}

verbose() {
	local msg="${1}"
	local level="${2:-1}"
	local i=0

	if [ ${level} -le ${VERBOSITY} ]; then
		while [ ${i} -lt ${level} ]; do
			printf "=" >&2
			i=$(( ${i} + 1 ))
		done
		echo "> ${msg}" >&2
	fi
}

vpnOff() {
	verbose "Disconnecting from VPN..."

	checkAnyconnectState "disconnected"
	${_ANYCONNECT} disconnect
}

vpnOn() {
	verbose "Connecting to VPN..."

	checkAnyconnectState "connected"
	getPasswordFromPasswordManager
	connectToVPN
}

###
### Main
###

setVariables

while getopts 'Vhlm:p:s:u:v' opt; do
	_FLAGS="${_FLAGS} ${opt}"
	case ${opt} in
		V)
			echo "${PROGNAME} Version ${VERSION}"
			exit 0
			# NOTREACHED
		;;
		h\?)
			usage
			exit 0
			# NOTREACHED
		;;
		l)
			_FORCE_LOGOUT=1
		;;
		m)
			_2FA="${OPTARG}"
		;;
		p)
			_PASSWORD_MANAGER="${OPTARG}"
		;;
		s)
			_VPN_SITE="${OPTARG}"
		;;
		u)
			_VPN_USER="${OPTARG}"
		;;
		v)
			VERBOSITY=$(( ${VERBOSITY} + 1 ))
		;;
		*)
			usage
			exit 1
			# NOTREACHED
		;;
	esac
done
shift $(($OPTIND - 1))

if [ $# -ne 1 ]; then
	usage
	exit 1
	# NOTREACHED
fi

checkEnv

case "${1}" in
	off)
		vpnOff
	;;
	on)
		vpnOn
	;;
	sites)
		listSites
	;;
	vars)
		dumpVars
	;;
	*)
		${_ANYCONNECT} "$@"
	;;
esac
